CGI: Fix cgi_interpreter configuration validation and execution

rousbound opened the pull request:
- Fixed cgi_interpreter configuration validation. Probably a typo on validate_configuration()?
- Also corrected get_executable function on unix systems. The program tried to execute files without shebang without assigning the configured cgi_interpreter.

rousbound commented:
Hi @DorianNiemiecSVRJS! First of all, congratulations on this great project!

I was wondering if I should’ve made an issue before the PR, but I thought this change was small enough to be proposed directly.

Still, let me know if you prefer this change to go through an issue.

rousbound commented:
I found one more problem:

The SERVER_NAME variable was being defined based on the "Host" header of the request. 

However, accordingly to RFC9110 the Host header might optionally have the port number and SERVER_NAME as specified in RFC3875 must be only servername|ipv4|ipv6.

Also, relying on client to set the SERVER_NAME doesn't seem ideal.

I propose the following order of sources to define SERVER_NAME:
- Trying to get the config hostname
- Getting ip address from socket_data

DorianNiemiecSVRJS commented on ferron-modules-builtin/src/optional/cgi.rs:
```diff
@@ -101,7 +101,7 @@ impl ModuleLoader for CgiModuleLoader {
       for entry in &entries.inner {
         if entry.values.first().is_some_and(|v| v.is_null())
-          || entry.values.get(1).is_some_and(|v| v.is_null() || v.is_string())
+          || entry.values.get(1).is_some_and(|v| v.is_null() || !v.is_string())
```
Thank you for catching the configuration validation bug!

I will cite from the Ferron documentation:
> - `cgi_interpreter <cgi_extension: string> <cgi_interpreter: string|null> [<cgi_interpreter_argument: string> ...]` (_cgi_ module)
>   - This directive specifies CGI script interpreters used by the CGI handler. If CGI interpreter is set to `#null`, the default interpreter settings will be disabled. This directive can be specified multiple times. Default: specified for `.pl`, `.py`, `.sh`, `.ksh`, `.csh`, `.rb` and `.php` extensions, and additionally `.exe`, `.bat` and `.vbs` extensions for Windows

I would suggest to change the part of this validation logic to this (ah, I forgot to invert the conditions in the code):

```rust
// ...snip...
if !entry.values.first().is_some_and(|v| v.is_string())
  || !entry.values.get(1).is_some_and(|v| v.is_null() || v.is_string())
{
  Err(anyhow::anyhow!("Invalid CGI extension interpreter specification"))?
}
// ...snip...
```

DorianNiemiecSVRJS commented on ferron-modules-builtin/src/optional/cgi.rs:
```diff
@@ -527,10 +528,12 @@ async fn execute_cgi_with_environment_variables(
   cgi_interpreters: HashMap<String, Vec<String>>,
   additional_environment_variables: HashMap<String, String>,
 ) -> Result<ResponseData, Box<dyn Error + Send + Sync>> {
+  println!("{:?}", request);
```
This `println!` statement seems to be a leftover after testing, so I would suggest to clean this up...

DorianNiemiecSVRJS commented on ferron-modules-builtin/src/optional/cgi.rs:
```diff
@@ -907,9 +911,22 @@ async fn get_executable(execute_pathbuf: &PathBuf) -> Result<Vec<String>, Box<dy
+
+  match open_file_result.1.freeze().as_ref() {
+    b"#!" => Ok(executable_params_vector),
+    _ => Err(anyhow::anyhow!("No shebang"))?
```
But what if a valid CGI program doesn't have a shebang (like it's the case in compiled CGI programs)?
I would suggest removing the shebang check (after checking if the CGI program is executable) for the `get_executable` function in case if the web server runs under Unix and -like.

DorianNiemiecSVRJS commented:
Thank you for the pull request!

I have reviewed it, and requested some changes...

rousbound commented:
Hi @DorianNiemiecSVRJS! 

Thanks for all your input!

I've made the changes you suggested:

- Inverted the conditions in the validation of the configuration
- Removed debugging leftovers
- Removed the shebang check

Now all tests are passing! (Didn't realized before there was a test suite)
